src/adaptation/Adaptation.jl

```diff
@@ -0,0 +1,40 @@
+module Adaptation
+
+import Base: string
+import LinearAlgebra, Statistics
+using ..AdvancedHMC: DEBUG
+
+abstract type AbstractAdapter end
+
+include("stepsize.jl")
+include("precond.jl")
+
+abstract type AbstractCompositeAdapter <: AbstractAdapter end
+
+# TODO: generalise this to a list of adapters
+struct NaiveCompAdapter <: AbstractCompositeAdapter
```

Comment:
Consider 

- removing `AbstractCompositeAdapter, and
- making `NaiveCompAdaptor` and `ThreePhaseAdaptor` direct subtypes of `AbstractAdapter`

Since `AbstractCompositeAdapter` is only used to define

- `getM⁻¹(ca::AbstractCompositeAdapter)`
- `getss(ca::AbstractCompositeAdapter)`
- `update(h::Hamiltonian, p::AbstractProposal, a::Adaptation.AbstractCompositeAdapter)`

These functions can be defined for `AbstractAdaptor` directly:

- `getM⁻¹(a::AbstractAdapter) =  a.var `
- `getss(a::AbstractAdapter) = a.ϵ`
- `update(h::Hamiltonian, p::AbstractProposal, a::Adaptation.AbstractCompositeAdapter)`

Then we can overload these functions for subtypes of `AbstractAdaptor`

- `getM⁻¹(a:: ThreePhaseAdapter) =  getM⁻¹(a.pc) `
- `getss(a:: ThreePhaseAdapter) = getss(a.ssa)`
- `getM⁻¹(dpc::UnitPreConditioner) = nothing` (perhaps return `1.0` instead of `nothing`)
- `getM⁻¹(dpc::DensePreConditioner) = dpc.covar`